### PR TITLE
ETCD Implementation of State Store

### PR DESCRIPTION
**Pull request by robertojrojas**
# Description

Adds [Etcd](https://etcd.io/) as a supported state component.

At the moment, this functionality needs integration testing. At the moment due , this is a bit tricky to the dependency between the **`dapr`** repo and this one. 

## Issue reference
Please reference the issue this PR will close: #41

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Extended the documentation

### CODE REVIEW

File: state/etcd/etcd.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+)
+
+const defaultOperationTimeout = time.Duration(10 * time.Second)
+const defaultSeparator = ","
+
+var errMissingEndpoints = errors.New("Endpoints are required")
+var errInvalidDialTimeout = errors.New("DialTimeout is invalid")
+
+// StateStore is a Etcd state store
+type StateStore struct {
```
Comment (yaron2) [inline state/etcd/etcd.go]: Hey, can you rename this to `ETCD`? I know in Redis we have it as StateStore, but that's going to change :)

File: state/etcd/etcd.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+	_, err := time.ParseDuration(configProps.DialTimeout)
+	if err != nil {
+		return errInvalidDialTimeout
+	}
+
+	return nil
+}
+
+// Get retrieves state from redis with a key
```
Comment (yaron2) [inline state/etcd/etcd.go]: Comment has redis in it

File: state/etcd/etcd.go (hunk 1 above)
Comment (robertojrojas) [reply state/etcd/etcd.go]: Just `ETCD` or `EtcdStateStore` ?

File: state/etcd/etcd.go (hunk 2 above)
Comment (robertojrojas) [reply state/etcd/etcd.go]: oops! :)

File: state/etcd/etcd.go (hunk 1 above)
Comment (yaron2) [reply state/etcd/etcd.go]: Just ETCD. Thanks!

Comment (robertojrojas) [reply state/etcd/etcd.go]: 👍

**Comment by yaron2** [comment]
Great!

Are you ready to merge yet?

**Comment by robertojrojas** [comment]
My integration test of POST and GET work fine. It's not perfect, but it's a good start. Good to merge!

**Comment by yaron2** [comment]
Whats missing from your point of view?

**Comment by robertojrojas** [comment]
TLS connectivity would be good to have. I was thinking on working on that later on. wdyt?

**Comment by yaron2** [comment]
TLS would be good, yes. That can be added later. I just wanted to see if there's anything which is a blocker. There isn't :)

Thanks for your contribution!

**Comment by yaron2** [comment]
TLS would be good, yes. That can be added later. I just wanted to see if there's anything which is a blocker. There isn't :)

Thanks for your contribution!

**Comment by robertojrojas** [comment]
You're welcome. This is a fun project to work on :)

**Comment by amanbha** [comment]
@robertojrojas  , @yaron2  Could you please make corresponding change in https://github.com/dapr/dapr/blob/master/pkg/components/state/loader.go to consume this.

**Comment by yaron2** [comment]
Aman I'll update dapr this week.

**Comment by robertojrojas** [comment]
@yaron2 If you want, I can PR the **loader** implementation I used to test this. Let me know.

**Comment by yaron2** [comment]
Sure, go ahead.

**Comment by yaron2** [comment]
Make sure you update the latest components-contrib in go.mod and run `go mod vendor` to pull the latest into the vendor folder.

**Pull request merged**